File: Maven_Accounting_Libs/mavenAccProductModule/src/main/java/com/krawler/hql/accounting/Inventory.java
/*
 * Copyright (C) 2012  Krawler Information Systems Pvt Ltd
 * All rights reserved.
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.krawler.hql.accounting;

import com.krawler.common.admin.Company;
import java.util.Date;

/**
 *
 * @author krawler-user
 */
public class Inventory {

    private String ID;
    private Product product;
    private double quantity;
    private UnitOfMeasure uom;
    private double baseuomquantity;
    private double baseuomrate;
    private double actquantity;
    private String description;
    private boolean carryIn;
    private boolean newInv;
    private boolean invrecord;
    private boolean defective;
    private Company company;
    private Date updateDate;
    private boolean deleted;
    private boolean openingInventory;
    private boolean leaseFlag;
    private boolean isconsignment;
    private boolean isjobworkorder;   
    private double consignuomquantity;
    private double venconsignuomquantity;
    private double productWeightPerStockUom;
    private double productWeightIncludingPakagingPerStockUom;
    private double productVolumePerStockUom;
    private double productVolumeIncludingPakagingPerStockUom;

    public double getProductWeightIncludingPakagingPerStockUom() {
        return productWeightIncludingPakagingPerStockUom;
    }

    public void setProductWeightIncludingPakagingPerStockUom(double productWeightIncludingPakagingPerStockUom) {
        this.productWeightIncludingPakagingPerStockUom = productWeightIncludingPakagingPerStockUom;
    }

    public double getProductWeightPerStockUom() {
        return productWeightPerStockUom;
    }

    public void setProductWeightPerStockUom(double productWeightPerStockUom) {
        this.productWeightPerStockUom = productWeightPerStockUom;
    }

    public double getConsignuomquantity() {
        return consignuomquantity;
    }

    public void setConsignuomquantity(double consignuomquantity) {
        this.consignuomquantity = consignuomquantity;
    }
    
    public boolean isIsconsignment() {
        return isconsignment;
    }

    public void setIsconsignment(boolean isconsignment) {
        this.isconsignment = isconsignment;
    }

    public Date getUpdateDate() {
        return updateDate;
    }

    public void setUpdateDate(Date updateDate) {
        this.updateDate = updateDate;
    }

    public String getID() {
        return ID;
    }

    public void setID(String ID) {
        this.ID = ID;
    }

    public boolean isCarryIn() {
        return carryIn;
    }

    public void setCarryIn(boolean carryIn) {
        this.carryIn = carryIn;
    }

    public boolean isNewInv() {
        return newInv;
    }

    public void setNewInv(boolean newInv) {
        this.newInv = newInv;
    }

    public boolean isInvrecord() {
        return invrecord;
    }

    public void setInvrecord(boolean invrecord) {
        this.invrecord = invrecord;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Product getProduct() {
        return product;
    }

    public void setProduct(Product product) {
        this.product = product;
    }

    public boolean isDefective() {
        return defective;
    }

    public void setDefective(boolean defective) {
        this.defective = defective;
    }

    public double getQuantity() {
        return quantity;
    }

    public void setQuantity(double quantity) {
        this.quantity = quantity;
    }

    public double getActquantity() {
        return actquantity;
    }

    public void setActquantity(double actquantity) {
        this.actquantity = actquantity;
    }

    public UnitOfMeasure getUom() {
        return uom;
    }

    public void setUom(UnitOfMeasure uom) {
        this.uom = uom;
    }

    public double getBaseuomquantity() {
        return baseuomquantity;
    }

    public void setBaseuomquantity(double baseuomquantity) {
        this.baseuomquantity = baseuomquantity;
    }

    public double getBaseuomrate() {
        return baseuomrate;
    }

    public void setBaseuomrate(double baseuomrate) {
        this.baseuomrate = baseuomrate;
    }

    public Company getCompany() {
        return company;
    }

    public void setCompany(Company company) {
        this.company = company;
    }

    public boolean isDeleted() {
        return deleted;
    }

    public void setDeleted(boolean deleted) {
        this.deleted = deleted;
    }

    public boolean isOpeningInventory() {
        return openingInventory;
    }

    public void setOpeningInventory(boolean openingInventory) {
        this.openingInventory = openingInventory;
    }

    public boolean isLeaseFlag() {
        return leaseFlag;
    }

    public void setLeaseFlag(boolean leaseFlag) {
        this.leaseFlag = leaseFlag;
    }

    public double getVenconsignuomquantity() {
        return venconsignuomquantity;
    }

    public void setVenconsignuomquantity(double venconsignuomquantity) {
        this.venconsignuomquantity = venconsignuomquantity;
    }

    public double getProductVolumePerStockUom() {
        return productVolumePerStockUom;
    }

    public void setProductVolumePerStockUom(double productVolumePerStockUom) {
        this.productVolumePerStockUom = productVolumePerStockUom;
    }

    public double getProductVolumeIncludingPakagingPerStockUom() {
        return productVolumeIncludingPakagingPerStockUom;
    }

    public void setProductVolumeIncludingPakagingPerStockUom(double productVolumeIncludingPakagingPerStockUom) {
        this.productVolumeIncludingPakagingPerStockUom = productVolumeIncludingPakagingPerStockUom;
    }

    public boolean isIsjobworkorder() {
        return isjobworkorder;
    }

    public void setIsjobworkorder(boolean isjobworkorder) {
        this.isjobworkorder = isjobworkorder;
    }
    }

File: Maven_Accounting/src/main/webapp/scripts/mainscripts/Inventory.js
/*
 * Copyright (C) 2012  Krawler Information Systems Pvt Ltd
 * All rights reserved.
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
Wtf.account.Inventory = function(config){
     Wtf.apply(this,{
         buttons: [{
                    text: WtfGlobal.getLocaleText("acc.het.108"),
                    scope: this,
                    //iconCls:'save',
                    handler: this.saveInventoryForm.createDelegate(this)
                  },{
                    text: WtfGlobal.getLocaleText("acc.CANCELBUTTON"),
                    scope: this,
                    handler: function(){ this.close();}
         }]
    },config);
    Wtf.account.Inventory.superclass.constructor.call(this, config);
     this.addEvents({
                'update':true
            });

}
Wtf.extend(  Wtf.account.Inventory, Wtf.Window, {
    onRender: function(config){
        Wtf.account.Inventory.superclass.onRender.call(this, config);
        this.createProductCombo();
        this.createFields();
        this.createForm();
        this.calculateAmount();
        this.add({
            region: 'north',
            height: 75,
            border: false,
            bodyStyle: 'background:white;border-bottom:1px solid #bfbfbf;',
            html: getTopHtml('Inventory','Inventory')
        },{
            region: 'center',
            border: false,
            baseCls:'bckgroundcolor',
            layout: 'fit',
            items:this.InventoryForm
        });
        this.rate.on('change',this.calculateAmount,this);//function(rate,val){alert(val)})
        this.quantity.on('change',this.calculateAmount,this);
    },
    calculateAmount:function(){
        this.amount.setValue(this.rate.getValue()*this.quantity.getValue());
    },
    createProductCombo:function(){
        this.CategoryRec = Wtf.data.Record.create ([
            {name:'categoryid'},
            {name:'categoryname'}
       ]);
       this.CategoryStore=new Wtf.data.Store({
           url: Wtf.req.account+'CompanyManager.jsp',
            baseParams:{
                 mode:71
            },
            reader: new Wtf.data.KwlJsonReader({
                root: "data"
            },this.CategoryRec)
       });
       this.CategoryStore.load();
       this.productRec=new Wtf.data.Record.create([
            {name:'productname'},
            {name:'categoryid'},
            {name:'categoryname'},
            {name:'specification'},
            {name:'productid'}
        ]);
        this.prodStore=new Wtf.data.Store({
            reader: new Wtf.data.KwlJsonReader({
                root: "data"
            },this.productRec),
            url: Wtf.req.account+'CompanyManager.jsp',
            baseParams:{
                mode:104
            }
        });
        this.specRec = Wtf.data.Record.create ([
            {name:'productname'},
            {name:'categoryid'},
            {name:'categoryname'},
            {name:'specification'},
            {name:'productid'}
        ]);
        this.specStore = new Wtf.data.Store({
            url:Wtf.req.account+'CompanyManager.jsp',
            baseParams:{
                mode:105
            },
        reader: new Wtf.data.KwlJsonReader({
            root: "data"
        },this.specRec)
    });
    this.specStore.load();
    },
    createFields:function(){
         this.productName= new Wtf.form.ComboBox({
            fieldLabel:WtfGlobal.getLocaleText("acc.productList.gridProduct"),
            hiddenName:'productid',
            store: this.prodStore,
            valueField:'productname',
            displayField:'productname',
            mode: 'local',
            allowBlank:false,
            triggerAction: 'all',
            editable : false,
            listeners:{
                scope:this,
                'select' :this.loadSpecificationStore.createDelegate(this)
            }
        });
        this.category= new Wtf.form.ComboBox({
            fieldLabel:WtfGlobal.getLocaleText("acc.cust.category"),
            hiddenName:'categoryid',
            store:this.CategoryStore,
            valueField:'categoryid',
            displayField:'categoryname',
            mode: 'local',
            allowBlank:false,
            triggerAction: 'all',
            editable : false,
            listeners:{
                scope:this,
                'select' :this.loadProdStore.createDelegate(this) 
            }
        });
        this.specification =new Wtf.form.ComboBox({
            fieldLabel:WtfGlobal.getLocaleText("acc.field.Specification"),
            hiddenName:'specification',
            name:'specification',
            store:this.specStore,
            valueField:'specification',
            displayField:'specification',
            mode: 'local',
            allowBlank:false,
            triggerAction:'all',
            forceSelection:true,
            editable : false
        });
        this.quantity= new Wtf.form.NumberField({
            fieldLabel: WtfGlobal.getLocaleText("acc.product.gridQty"),
            name: 'quantity',
            allowBlank:false,
            maxLength:4,
            id:'quantity',
            xtype:'numberfield',
            vtype:'alphanum'
        });
        this.rate= new Wtf.form.NumberField({
            fieldLabel: WtfGlobal.getLocaleText("acc.field.Rate"),
            name: 'rate',
            allowBlank:false,            
            id:'rate',
            xtype:'numberfield',
            vtype:'alphanum',
            maxLength:8,
            decimalPrecision:2
        });
        this.amount= new Wtf.form.NumberField({
            fieldLabel: WtfGlobal.getLocaleText("acc.dnList.gridAmt"),
            name: 'amount',
            id:'amount',
            readOnly :true,
            xtype:'numberfield',
            vtype:'alphanum'
        });
        this.remark= new Wtf.form.TextField({
            fieldLabel: WtfGlobal.getLocaleText("acc.invoice.gridRemark"),
            name: 'description'
        });
    },
    createForm:function(){
        this.InventoryForm=new Wtf.form.FormPanel({
            region:'center',
            width:'400',
            height:'250',
            border:false,
            bodyStyle: "background: transparent;",
            style: "background: transparent;padding-left: 35px;padding-top: 20px;padding-right: 30px;",
            defaultType: 'textfield',
            defaults:{
                width:200
            },
            items:[{xtype:'hidden',name:'inventoryid'},this.category,this.productName,this.specification,this.quantity,this.rate,this.amount,this.remark]
        }); 

    },
    saveInventoryForm:function(){
         if(!this.InventoryForm.getForm().isValid()){
                 WtfComMsgBox(2,2);
         }else{
                var rec=this.InventoryForm.getForm().getValues();
                rec.productid=this.specStore.getAt(this.specStore.find('specification',this.specification.getValue())).data['productid'];
                rec.mode=41;
                Wtf.Ajax.requestEx({
                url: Wtf.req.account+'CompanyManager.jsp',
                params: rec
           },this,this.genSuccessResponse,this.genFailureResponse);
         }

        },
    genSuccessResponse:function(response){
        WtfComMsgBox([WtfGlobal.getLocaleText("acc.invoice.inventory"),response.msg],response.success*2+1);
        if(response.success) this.fireEvent('update',this);
        this.close();
    }, 
    genFailureResponse:function(response){
        var msg=WtfGlobal.getLocaleText("acc.field.FailedtomakeconnectionwithWebServer");
        if(response.msg)msg=response.msg;
        WtfComMsgBox([WtfGlobal.getLocaleText("acc.common.alert"),msg],2);
        this.close();
    },
    loadSpecificationStore:function(v){
         this.specStore.load({
            params:{
                productid:v.getValue(),
                productcategoryid:this.category.getValue()
            }});
        this.specification.setValue("");
    },
    loadProdStore:function(v){
        this.prodStore.load({
            params:{
                productcategoryid:v.getValue()
            }
        });
        this.specification.setValue("");
        this.productName.setValue("");
    }
}); 